pkg/p2p/kadcast/network.go

```diff
@@ -61,5 +63,61 @@ func sendUDPPacket(netw string, addr net.UDPAddr, payload []byte) {
 	if err != nil {
 		log.WithError(err).Warn("Error while writting to the filedescriptor.")
 		return
-	} 
-}
\ No newline at end of file
+	}
+}
+
+// StartTCPListener listens infinitely for TCP packet arrivals and
+// executes it's processing inside a gorutine by sending
+// the packets to the circularQueue.
+func StartTCPListener(netw string, queue *ring.Buffer, MyPeerInfo Peer) {
+	lAddr := getLocalTCPAddress()
+	// Set listening port.
+	lAddr.Port = int(MyPeerInfo.port)
+PacketConnCreation:
+	// listen to incoming TCP packets
+	listener, err := net.ListenTCP(netw, &lAddr)
+	if err != nil {
+		log.Panic(err)
+	}
+	// Set initial deadline.
+	listener.SetDeadline(time.Now().Add(time.Minute))
+
+	// Instanciate the buffer
+	buffer := make([]byte, 5000000)
+	for {
+		// Read TCP packet.
+		pc, err := listener.AcceptTCP()
+		uAddr := pc.RemoteAddr()
+		byteNum, err := io.ReadFull(pc, buffer)
+		if err != nil {
+			log.WithError(err).Warn("Error on packet read")
+			pc.Close()
+			goto PacketConnCreation
+		}
+		// Set a new deadline for the connection.
+		pc.SetDeadline(time.Now().Add(5 * time.Minute))
+		// Serialize the packet.
+		encodedPack := encodeReadTCPPacket(uint16(byteNum), uAddr, buffer[0:byteNum])
+		// Send the packet to the Consumer putting it on the queue.
+		queue.Put(encodedPack)
+		// Empty the buffer to read again.
+		buffer = make([]byte, 5000000)
+	}
+}
+
+// Opens a TCP connection with the peer sent on the params and transmits
+// a stream of bytes. Once transmited, closes the connection.
+func sendTCPStream(addr net.UDPAddr, payload []byte) {
+	conn, err := net.Dial("tcp", string(addr.IP)+":"+string(addr.Port))
+	if err != nil {
+		log.WithError(err).Warn("Could not stablish a connection with the dest Peer.")
+		return
+	}
+	defer conn.Close()
+	// Write our message to the connection.
+	_, err = conn.Write(payload)
```

Comment:
Is there a reason we don't have `WriteDeadline` prior to this Write attempt?

Also, it could be good to call `conn.Write` in a separate goroutine (non-blocking write), otherwise blocking here would freeze ProcessPacket and eventually `ringbuffer` would start dropping messages

Reply:
I agree that it can be a good idea to maki it non-blocking. For the deadline stuff, I opened an issue since on UDP they cause crashes at some point. But they didn't solve anything..

Reply:
If `SetDeadline` fails on UDP, we can have a separate issue to investigate it. To what I have tested with SetWriteDeadline on UDP conn hasn't given any issues so far. Is there a golang issue referring to such problem?